scheds/rust/scx_rusty/src/main.rs

```diff
@@ -446,17 +452,41 @@ impl<'a> Scheduler<'a> {
         lb_stats: &[NumaStat],
     ) {
         let stat = |idx| bpf_stats[idx as usize];
-        let total = stat(bpf_intf::stat_idx_RUSTY_STAT_WAKE_SYNC)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_SYNC_PREV_IDLE)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_PREV_IDLE)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_GREEDY_IDLE)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_PINNED)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_DIRECT_DISPATCH)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_DIRECT_GREEDY)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_DIRECT_GREEDY_FAR)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_DSQ_DISPATCH)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_GREEDY_LOCAL)
-            + stat(bpf_intf::stat_idx_RUSTY_STAT_GREEDY_XNUMA);
+
+        let wsync = stat(bpf_intf::stat_idx_RUSTY_STAT_WAKE_SYNC);
+        let wsync_prev_idle = stat(bpf_intf::stat_idx_RUSTY_STAT_SYNC_PREV_IDLE);
+        let prev_idle = stat(bpf_intf::stat_idx_RUSTY_STAT_PREV_IDLE);
+        let greedy_idle = stat(bpf_intf::stat_idx_RUSTY_STAT_GREEDY_IDLE);
+        let pinned = stat(bpf_intf::stat_idx_RUSTY_STAT_PINNED);
+        let direct_dispatch = stat(bpf_intf::stat_idx_RUSTY_STAT_DIRECT_DISPATCH);
+        let direct_greedy = stat(bpf_intf::stat_idx_RUSTY_STAT_DIRECT_GREEDY);
+        let direct_greedy_far = stat(bpf_intf::stat_idx_RUSTY_STAT_DIRECT_GREEDY_FAR);
+        let dsq = stat(bpf_intf::stat_idx_RUSTY_STAT_DSQ_DISPATCH);
+        let greedy_local = stat(bpf_intf::stat_idx_RUSTY_STAT_GREEDY_LOCAL);
+        let greedy_xnuma = stat(bpf_intf::stat_idx_RUSTY_STAT_GREEDY_XNUMA);
+        let total = wsync
+            + wsync_prev_idle
+            + prev_idle
+            + greedy_idle
+            + pinned
+            + direct_dispatch
+            + direct_greedy
+            + direct_greedy_far
+            + dsq
+            + greedy_local
+            + greedy_xnuma;
+
+        counter!("dispatched_tasks_count", "type" => "wsync_prev_idle").increment(wsync_prev_idle);
+        counter!("dispatched_tasks_count", "type" => "wsync").increment(wsync);
+        counter!("dispatched_tasks_count", "type" => "prev_idle").increment(prev_idle);
+        counter!("dispatched_tasks_count", "type" => "greedy_idle").increment(greedy_idle);
+        counter!("dispatched_tasks_count", "type" => "pinned").increment(pinned);
+        counter!("dispatched_tasks_count", "type" => "direct_dispatch").increment(direct_dispatch);
+        counter!("dispatched_tasks_count", "type" => "direct_greedy").increment(direct_greedy);
+        counter!("dispatched_tasks_count", "type" => "direct_greedy_far").increment(direct_greedy_far);
+        counter!("dispatched_tasks_count", "type" => "dsq").increment(dsq);
+        counter!("dispatched_tasks_count", "type" => "greedy_local").increment(greedy_local);
+        counter!("dispatched_tasks_count", "type" => "greedy_xnuma").increment(greedy_xnuma);
```

Comment:
This registers a new counter on each `report()`. I would think we want to register the counters ahead of time and only set their absolute value here, no?

Reply:
Let me look into that and see if we'd benefit from instantiating them once and reusing them. I was under the impression that the library authors want this pattern and have optimized the implementation to avoid the perceived overhead of registering new counters on each call.

Reply:
@dschatzberg I created a `Metrics` struct to hold the metrics and made it a member of the Scheduler struct. We now only register the metrics once.